File: ts/src/meshtrade/type/v1/date.test.ts
import {
  newDate,
  newDateFromJsDate,
  isValid,
  isComplete,
  dateToJsDate,
  dateToString,
} from "./date";
import { DateSchema, Date as ProtoDate } from "./date_pb";
import { create } from "@bufbuild/protobuf";

describe("newDate", () => {
  test("valid complete date", () => {
    const result = newDate(2023, 12, 25);
    expect(result.year).toBe(2023);
    expect(result.month).toBe(12);
    expect(result.day).toBe(25);
  });

  test("valid leap year February 29", () => {
    const result = newDate(2024, 2, 29);
    expect(result.year).toBe(2024);
    expect(result.month).toBe(2);
    expect(result.day).toBe(29);
  });

  test("invalid year zero", () => {
    expect(() => newDate(0, 12, 25)).toThrow("Year must be between 1 and 9999");
  });

  test("invalid year too high", () => {
    expect(() => newDate(10000, 1, 1)).toThrow(
      "Year must be between 1 and 9999"
    );
  });

  test("invalid month zero", () => {
    expect(() => newDate(2023, 0, 25)).toThrow(
      "Month must be between 1 and 12"
    );
  });

  test("invalid month too high", () => {
    expect(() => newDate(2023, 13, 1)).toThrow(
      "Month must be between 1 and 12"
    );
  });

  test("invalid day zero", () => {
    expect(() => newDate(2023, 1, 0)).toThrow("Day must be between 1 and 31");
  });

  test("invalid day too high", () => {
    expect(() => newDate(2023, 1, 32)).toThrow("Day must be between 1 and 31");
  });

  test("invalid date February 30", () => {
    expect(() => newDate(2023, 2, 30)).toThrow("Invalid date");
  });

  test("invalid leap year February 29", () => {
    expect(() => newDate(2023, 2, 29)).toThrow("Invalid date");
  });
});

describe("newDateFromJsDate", () => {
  const testCases: Array<[string, Date, number, number, number]> = [
    ["standard date", new Date(2023, 11, 25), 2023, 12, 25], // JS months are 0-indexed
    ["leap year date", new Date(2024, 1, 29), 2024, 2, 29],
    ["first day of year", new Date(2023, 0, 1), 2023, 1, 1],
    ["last day of year", new Date(2023, 11, 31), 2023, 12, 31],
  ];

  test.each(testCases)(
    "%s",
    (_, jsDate, expectedYear, expectedMonth, expectedDay) => {
      const result = newDateFromJsDate(jsDate);
      expect(result.year).toBe(expectedYear);
      expect(result.month).toBe(expectedMonth);
      expect(result.day).toBe(expectedDay);
    }
  );
});

describe("isValid", () => {
  const testCases: Array<[string, ProtoDate | undefined, boolean]> = [
    ["undefined date", undefined, false],
    ["valid complete date", newDate(2023, 12, 25), true],
    [
      "invalid year zero",
      create(DateSchema, { year: 0, month: 12, day: 25 }),
      false,
    ],
    [
      "invalid year too high",
      create(DateSchema, { year: 10000, month: 1, day: 1 }),
      false,
    ],
    [
      "invalid month zero",
      create(DateSchema, { year: 2023, month: 0, day: 25 }),
      false,
    ],
    [
      "invalid month too high",
      create(DateSchema, { year: 2023, month: 13, day: 1 }),
      false,
    ],
    [
      "invalid day zero",
      create(DateSchema, { year: 2023, month: 1, day: 0 }),
      false,
    ],
    [
      "invalid day too high",
      create(DateSchema, { year: 2023, month: 1, day: 32 }),
      false,
    ],
  ];

  test.each(testCases)("%s", (_, date, expected) => {
    expect(isValid(date)).toBe(expected);
  });
});

describe("isComplete", () => {
  const testCases: Array<[string, ProtoDate | undefined, boolean]> = [
    ["undefined date", undefined, false],
    ["complete date", newDate(2023, 12, 25), true],
    [
      "incomplete - year zero",
      create(DateSchema, { year: 0, month: 12, day: 25 }),
      false,
    ],
    [
      "incomplete - month zero",
      create(DateSchema, { year: 2023, month: 0, day: 25 }),
      false,
    ],
    [
      "incomplete - day zero",
      create(DateSchema, { year: 2023, month: 12, day: 0 }),
      false,
    ],
    ["zero date", create(DateSchema, { year: 0, month: 0, day: 0 }), false],
  ];

  test.each(testCases)("%s", (_, date, expected) => {
    expect(isComplete(date)).toBe(expected);
  });
});

describe("dateToJsDate", () => {
  test("valid complete date", () => {
    const date = newDate(2023, 12, 25);
    const result = dateToJsDate(date);
    expect(result).toEqual(new Date(2023, 11, 25)); // JS months are 0-indexed
  });

  test("leap year date", () => {
    const date = newDate(2024, 2, 29);
    const result = dateToJsDate(date);
    expect(result).toEqual(new Date(2024, 1, 29));
  });

  test("invalid date throws error", () => {
    const date = create(DateSchema, { year: 2023, month: 2, day: 30 });
    expect(() => dateToJsDate(date)).toThrow();
  });
});

describe("dateToString", () => {
  const testCases: Array<[string, ProtoDate | undefined, string]> = [
    ["undefined date", undefined, "<undefined>"],
    ["complete date", newDate(2023, 12, 25), "2023-12-25"],
    [
      "invalid date",
      create(DateSchema, { year: 0, month: 12, day: 25 }),
      "Date{year=0, month=12, day=25} [INVALID]",
    ],
    [
      "invalid date February 30",
      create(DateSchema, { year: 2023, month: 2, day: 30 }),
      "Date{year=2023, month=2, day=30} [INVALID]",
    ],
  ];

  test.each(testCases)("%s", (_, date, expected) => {
    expect(dateToString(date)).toBe(expected);
  });
});

File: go/type/v1/date_test.go
package typev1

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestNewDate(t *testing.T) {
	tests := []struct {
		name      string
		year      int32
		month     int32
		day       int32
		wantErr   bool
		wantDate  *Date
		errSubstr string
	}{
		{
			name:     "valid complete date",
			year:     2023,
			month:    12,
			day:      25,
			wantErr:  false,
			wantDate: &Date{Year: 2023, Month: 12, Day: 25},
		},
		{
			name:      "invalid year zero",
			year:      0,
			month:     12,
			day:       25,
			wantErr:   true,
			errSubstr: "year must be between 1 and 9999",
		},
		{
			name:      "invalid year too high",
			year:      10000,
			month:     1,
			day:       1,
			wantErr:   true,
			errSubstr: "year must be between 1 and 9999",
		},
		{
			name:      "invalid month zero",
			year:      2023,
			month:     0,
			day:       1,
			wantErr:   true,
			errSubstr: "month must be between 1 and 12",
		},
		{
			name:      "invalid month too high",
			year:      2023,
			month:     13,
			day:       1,
			wantErr:   true,
			errSubstr: "month must be between 1 and 12",
		},
		{
			name:      "invalid day zero",
			year:      2023,
			month:     1,
			day:       0,
			wantErr:   true,
			errSubstr: "day must be between 1 and 31",
		},
		{
			name:      "invalid day too high",
			year:      2023,
			month:     1,
			day:       32,
			wantErr:   true,
			errSubstr: "day must be between 1 and 31",
		},
		{
			name:      "invalid date February 30",
			year:      2023,
			month:     2,
			day:       30,
			wantErr:   true,
			errSubstr: "invalid date",
		},
		{
			name:      "invalid leap year February 29 in non-leap year",
			year:      2023,
			month:     2,
			day:       29,
			wantErr:   true,
			errSubstr: "invalid date",
		},
		{
			name:     "valid leap year February 29",
			year:     2024,
			month:    2,
			day:      29,
			wantErr:  false,
			wantDate: &Date{Year: 2024, Month: 2, Day: 29},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := NewDate(tt.year, tt.month, tt.day)
			if tt.wantErr {
				require.Error(t, err)
				assert.Contains(t, err.Error(), tt.errSubstr)
				assert.Nil(t, got)
			} else {
				require.NoError(t, err)
				assert.Equal(t, tt.wantDate, got)
			}
		})
	}
}

func TestDate_ToTime(t *testing.T) {
	tests := []struct {
		name        string
		date        *Date
		wantPanic   bool
		panicSubstr string
		wantTime    time.Time
	}{
		{
			name:      "valid complete date",
			date:      &Date{Year: 2023, Month: 12, Day: 25},
			wantPanic: false,
			wantTime:  time.Date(2023, 12, 25, 0, 0, 0, 0, time.UTC),
		},
		{
			name:      "nil date",
			date:      nil,
			wantPanic: false,
			wantTime:  time.Time{},
		},
		{
			name:        "invalid date",
			date:        &Date{Year: 2023, Month: 2, Day: 30},
			wantPanic:   true,
			panicSubstr: "date is invalid",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.wantPanic {
				assert.PanicsWithValue(t, tt.panicSubstr, func() {
					tt.date.ToTime()
				})
			} else {
				got := tt.date.ToTime()
				assert.Equal(t, tt.wantTime, got)
			}
		})
	}
}

func TestDate_IsValid(t *testing.T) {
	tests := []struct {
		name string
		date *Date
		want bool
	}{
		{"nil date", nil, false},
		{"valid complete date", &Date{Year: 2023, Month: 12, Day: 25}, true},
		{"invalid year zero", &Date{Year: 0, Month: 12, Day: 25}, false},
		{"invalid year too high", &Date{Year: 10000, Month: 1, Day: 1}, false},
		{"invalid month zero", &Date{Year: 2023, Month: 0, Day: 25}, false},
		{"invalid month too high", &Date{Year: 2023, Month: 13, Day: 1}, false},
		{"invalid day zero", &Date{Year: 2023, Month: 1, Day: 0}, false},
		{"invalid day too high", &Date{Year: 2023, Month: 1, Day: 32}, false},
		{"invalid date February 30", &Date{Year: 2023, Month: 2, Day: 30}, false},
		{"valid leap year", &Date{Year: 2024, Month: 2, Day: 29}, true},
		{"invalid leap year", &Date{Year: 2023, Month: 2, Day: 29}, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.date.IsValid()
			assert.Equal(t, tt.want, got)
		})
	}
}

func TestDate_IsComplete(t *testing.T) {
	tests := []struct {
		name string
		date *Date
		want bool
	}{
		{"nil date", nil, false},
		{"complete date", &Date{Year: 2023, Month: 12, Day: 25}, true},
		{"incomplete - year zero", &Date{Year: 0, Month: 12, Day: 25}, false},
		{"incomplete - month zero", &Date{Year: 2023, Month: 0, Day: 25}, false},
		{"incomplete - day zero", &Date{Year: 2023, Month: 12, Day: 0}, false},
		{"zero date", &Date{Year: 0, Month: 0, Day: 0}, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.date.IsComplete()
			assert.Equal(t, tt.want, got)
		})
	}
}

func TestNewDateFromTime(t *testing.T) {
	tests := []struct {
		name     string
		timeVal  time.Time
		wantDate *Date
	}{
		{
			name:     "standard date",
			timeVal:  time.Date(2023, 12, 25, 15, 30, 45, 123456789, time.UTC),
			wantDate: &Date{Year: 2023, Month: 12, Day: 25},
		},
		{
			name:     "leap year date",
			timeVal:  time.Date(2024, 2, 29, 0, 0, 0, 0, time.UTC),
			wantDate: &Date{Year: 2024, Month: 2, Day: 29},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := NewDateFromTime(tt.timeVal)
			assert.Equal(t, tt.wantDate, got)
		})
	}
}
